datafusion/core/src/physical_optimizer/replace_with_order_preserving_variants.rs

```diff
@@ -15,136 +15,256 @@
-    plan.maintains_input_order().iter().any(|flag| *flag)
-}
+impl OrderPreservationContext {
+    /// Creates a "default" order-preservation context.
+    pub fn new(plan: Arc<dyn ExecutionPlan>) -> Self {
+        let length = plan.children().len();
+        OrderPreservationContext {
+            plan,
+            ordering_onwards: vec![None; length],
+        }
+    }
 
-/// Check the children nodes of a `SortExec` until ordering is lost (e.g. until
-/// another `SortExec` or a `CoalescePartitionsExec` which doesn't maintain ordering)
-/// and replace `RepartitionExec`s that do not maintain ordering (e.g. those whose
-/// input partition counts are larger than unity) with `SortPreservingRepartitionExec`s.
-/// Note that doing this may render the `SortExec` in question unneccessary, which will
-/// be removed later on.
-///
-/// For example, we transform the plan below
-/// "FilterExec: c@2 > 3",
-/// "  RepartitionExec: partitioning=Hash(\[b@0], 16), input_partitions=16",
-/// "    RepartitionExec: partitioning=Hash(\[a@0], 16), input_partitions=1",
-/// "      MemoryExec: partitions=1, partition_sizes=\[(<depends_on_batch_size>)], output_ordering: \[PhysicalSortExpr { expr: Column { name: \"a\", index: 0 }, options: SortOptions { descending: false, nulls_first: false } }]",
-/// into
-/// "FilterExec: c@2 > 3",
-/// "  SortPreservingRepartitionExec: partitioning=Hash(\[b@0], 16), input_partitions=16",
-/// "    RepartitionExec: partitioning=Hash(\[a@0], 16), input_partitions=1",
-/// "      MemoryExec: partitions=1, partition_sizes=\[<depends_on_batch_size>], output_ordering: \[PhysicalSortExpr { expr: Column { name: \"a\", index: 0 }, options: SortOptions { descending: false, nulls_first: false } }]",
-/// where the `FilterExec` in the latter has output ordering `a ASC`. This ordering will
-/// potentially remove a `SortExec` at the top of `FilterExec`. If this doesn't help remove
-/// a `SortExec`, the old version is used.
-fn replace_sort_children(
-    plan: &Arc<dyn ExecutionPlan>,
-) -> Result<Arc<dyn ExecutionPlan>> {
-    if plan.children().is_empty() {
-        return Ok(plan.clone());
+    /// Creates a new order-preservation context from those of children nodes.
+    pub fn new_from_children_nodes(
+        children_nodes: Vec<OrderPreservationContext>,
+        parent_plan: Arc<dyn ExecutionPlan>,
+    ) -> Result<Self> {
+        let children_plans = children_nodes
+            .iter()
+            .map(|item| item.plan.clone())
+            .collect();
+        let ordering_onwards = children_nodes
+            .into_iter()
+            .enumerate()
+            .map(|(idx, item)| {
+                // Leaves of the `coalesce_onwards` tree are `CoalescePartitionsExec`
+                // operators. This tree collects all the intermediate executors that
+                // maintain a single partition. If we just saw a `CoalescePartitionsExec`
+                // operator, we reset the tree and start accumulating.
```

Comment:
This comment is outdated and confusing, let's remove and replace with an appropriate one.